### PR TITLE
perf(training): optimize for RTX 3070Ti Laptop GPU

### LINKED ISSUE #42: perf(training): optimize training pipeline for NVIDIA RTX 3070Ti Laptop GPU
## Context
Training currently takes too long on the developer's local machine. The goal is to optimize the training pipeline to utilize the NVIDIA RTX 3070Ti **Laptop GPU** more efficiently.

## Hardware Specifications
- **GPU:** NVIDIA RTX 3070Ti Laptop GPU
- **VRAM:** 8GB
- **CUDA Cores:** 6144
- **Features:** Mixed precision support (FP16), Tensor Cores
- **Note:** Laptop variant — operates under lower TDP/power limits than desktop counterpart; power limits may cap GPU utilization under sustained load

## Suggested Investigation Areas

1. **Mixed Precision Training**
   - Implement torch.cuda.amp for automatic mixed precision (AMP)
   - Use FP16/BF16 to reduce memory footprint and leverage tensor core operations
   - Minimal code changes, significant speedup potential

2. **CUDA Optimization**
   - Enable cudnn.benchmark: `torch.backends.cudnn.benchmark = True` for fixed-size inputs
   - Verify CUDA is actually being used (check nvidia-smi during training)
   - Check laptop power mode (prefer "High Performance" or disable power-saving GPU profiles)

3. **Batch Size Optimization**
   - Current style_glyphs shape: [B, 10, 1, 128, 128] is memory intensive
   - Investigate if VRAM allows for larger batch sizes
   - Balance between memory usage and training efficiency

4. **Data Loading Pipeline**
   - Use DataLoader with num_workers (start with 4)
   - Enable pin_memory=True for faster host-to-device transfers
   - Profile data loading time as % of total epoch time

5. **Profiling & Diagnosis**
   - Use torch.profiler to identify actual bottleneck (data loading vs forward/backward pass)
   - Check if memory is the constraint (consider gradient checkpointing if needed)
   - Monitor GPU utilization and memory usage during training (`nvidia-smi dmon`)

## Acceptance Criteria
- [ ] Training time per epoch measurably reduced on RTX 3070Ti Laptop GPU hardware
- [ ] Performance improvement validated with before/after benchmarks
- [ ] Changes documented in training guide

### PR DESCRIPTION
**Pull request by FjoNef**
Implements AMP, cudnn.benchmark, DataLoader pin_memory/persistent_workers, epoch timing, and training guide. Closes #42

### CODE REVIEW

**Comment by FjoNef** [comment]
## QA Review — PR #43: AMP + RTX 3070Ti Training Optimizations

**Reviewer:** Saito (QA)  
**Verdict: CHANGES REQUESTED** — Two actionable issues; the rest of the implementation is solid.

---

### ✅ What's Good

**AMP implementation (train.py):**
- Separate `GradScaler` instances per optimizer (`scaler_g`, `scaler_d`) — correct for independent GAN backward passes.
- `autocast` is scoped correctly: forward pass + loss **inside**, `backward()` **outside**. Verified for both D and G steps.
- `GradScaler(enabled=use_amp)` / `autocast(enabled=use_amp)` — both are no-ops when CUDA is unavailable. CPU training path is clean.
- No gradient clipping present → `scaler.unscale_()` not required.
- `torch.backends.cudnn.benchmark = True` is only set inside `if torch.cuda.is_available():` — no CPU bleed.

**ONNX export isolation:**
- `export_onnx.py` is completely isolated: `device = torch.device('cpu')` hardcoded, models loaded fresh from checkpoint with `.eval()`, wrapped in `torch.no_grad()`. No `autocast` or `GradScaler` anywhere in the export path. ✅

**TRAINING.md:**
- Excellent coverage: prerequisites, power mode (Windows + Linux), `nvidia-smi dmon` commands, epoch timing, AMP rationale, cuDNN benchmark explanation, batch size table with VRAM estimates, performance comparison table.
- The note that `persistent_workers` is especially impactful on Windows (expensive process spawning) is a valuable practical detail.

**Existing regression tests:**
- All 9 tests (Bug 1 style conditioning + Bug 2 lambda_l1/feature-matching) remain valid and unaffected by the AMP change.

---

### ❌ Required Changes

**1. `persistent_workers=True` has no explicit guard against `num_workers=0`**

Current code:
```python
use_persistent_workers = True  # unconditional
train_loader = DataLoader(
    ...,
    num_workers=min(4, os.cpu_count() or 1),
    persistent_workers=use_persistent_workers,
)
```

`persistent_workers=True` with `num_workers=0` raises `ValueError` at DataLoader construction. Right now this is safe because `min(4, os.cpu_count() or 1)` always yields ≥ 1, but the coupling is implicit. A future developer debugging with `num_workers=0` will hit this instantly with a confusing error.

**Fix — tie the flag to the actual worker count:**
```python
num_workers = min(4, os.cpu_count() or 1)
train_loader = DataLoader(
    ...,
    num_workers=num_workers,
    persistent_workers=num_workers > 0,  # explicit guard
)
```

**2. No AMP smoke test — a GPU-specific PR needs at least one new test**

The 9 existing tests validate style conditioning and loss weights; none of them exercise the AMP forward pass. For a performance PR whose central change is `autocast` + `GradScaler`, there should be a test that verifies:
- A single training step with `use_amp=True` (or the CPU-safe `enabled=False` equivalent) completes without NaN/Inf losses.
- Scaler state advances (i.e., `scaler.get_scale()` doesn't drop to 0 from an immediate inf skip).

Suggested test file: `tests/test_amp_training_step.py`. It can run on CPU using `GradScaler(enabled=False)` + `autocast(enabled=False)` so it passes in CI without a GPU. The absence of this test means the AMP wiring is only validated manually on the target hardware.

---

### ⚠️ Minor Notes (non-blocking, good to fix before merge)

**3. Deprecated import path** (`torch.cuda.amp`):
```python
from torch.cuda.amp import GradScaler, autocast  # deprecated in PyTorch 2.0+
```
PyTorch 2.4+ emits `FutureWarning` for this. Prefer:
```python
from torch.amp import GradScaler, autocast
# and pass device_type='cuda' to autocast:
with autocast(device_type='cuda', enabled=use_amp):
```

**4. TRAINING.md code snippet shows `pin_memory=True` unconditionally**, but the actual code is:
```python
use_pin_memory = device.type == 'cuda'
```
Minor documentation drift — update the snippet to match.

---

### Summary

The core AMP implementation is correct and the ONNX export is cleanly isolated — no correctness bugs. Issues 1 and 2 above are the blockers: issue 1 is a fragile pattern that will bite the next person who sets `num_workers=0`; issue 2 is a meaningful test coverage gap for the PR's primary change. Both are small fixes.

**Comment by FjoNef** [comment]
Closing in favour of correctly-branched PR from squad/42-training-perf -> dev
